### ISSUE TITLE
Help / Documentation wanted for how to integrate third-party lib

### ISSUE THREAD

**Issue by trittweiler** [mesonbuild/meson#4872]
I would really appreciate some help on how to integrate a third-party lib. The source of the third-party lib is part of my project's source repository as a subproject.

I do not want to write a meson build file on behalf of the third-party. That's not really feasible for big third-party projects.

I can build the third-party project by having `custom_target()` invoke a build script. Then the library is built into some sub-directory. Unfortunately: the `output:` of `custom_target()` cannot contain a path separator. And also `static_library()` does not provide a means to specify a directory (e.g. `static_library(foo, [], :link_depends foo_target, directory: 'some/sub/directory/')` would look for `libfoo.a` in current_build_root() / 'some/sub/directory/')

Then I tried to not just build the third-party project, but actually install it under its build root because it comes with pkg-config definition files. But then there's no way to make `dependency(..., method: 'pkg-config')` to look for the dependency in that install directory within the current build root.

I would really appreciate some information on best practises here. 

Thanks in advance!

**Comment by xclaesse** [comment]
I made a PR to support that: https://github.com/mesonbuild/meson/pull/4321.

**Comment by xclaesse** [comment]
Sadly it doesn't have the choice because you need to build and install the dep to be able to get its pkg-config file in the configure. Unless you have better idea?

**Comment by xclaesse** [comment]
With my PR the third party takes precedence over system dep. If not the that's a bug.

**Comment by trittweiler** [comment]
> Sadly it doesn't have the choice because you need to build and install the dep to be able to get its pkg-config file in the configure. Unless you have better idea?

I wasn't aware that `dependency(...)` resolves the dependencies at configuration time. It makes sense, of course. It just wasn't part of my mental model. Hard to see a way around that with the current modus operandi.

**Comment by trittweiler** [comment]
> With my PR the third party takes precedence over system dep. If not the that's a bug.
> 
> Well, depends on how you call your third party subproject. If you set it as fallbacks to dependency() then system takes precedence, but you can call subproject() directly instead to and that won't lookup for system dep.

The `External-Project-module.md` of commit c5ae61f72c84 says:

_This module allows building code that uses build systems other than Meson. This module is intended to be used to build Autotools or CMake subprojects as fallback if the dependency couldn't be found on the system (e.g. too old distro version)._

Which I interpreted as it would look inside the system directories first. You may want to adapt that paragraph to prevent confusion.

**Comment by xclaesse** [comment]
Right, that doc could be improved, good catch. Note you don't even have to make it a subproject, it can even be just a subdir in your main project.
